### PR TITLE
Dynamic management of running plugins and framework rework

### PR DESCRIPTION
**Pull request by phoracek**
In this branch I'm decoupling some components of DPM in order to make it possible to dynamically add and remove plugins (when a new resource device appears). This change also should make architecture clearer and and also easier to create new plugins.

arch

TODO:
- [x] make plugin registration dynamic
- [x] expose Run and Stop to plugins
- [x] make sure that stop signal is handled correctly

### CODE REVIEW

**Comment by phoracek** [comment]
Ok, it should be ready for a review. Sorry about the amount of patches, the best way to review it would probably be checking dpm/ modules and their documentation in the current form. It builds, network/bridge plugins works. Will try to test kvm and vfio plugins as well.

With this change it is presentable for external users. We should consider moving it into a separate repo, so it is more appealing and "minimal".

File: pkg/dpm/lister.go [hunk 1]
```diff
@@ -1,9 +1,26 @@
 package dpm
 
-type DeviceMap map[string][]string
+// ResourceLastNamesList contains last names of discovered resources. This type is used by Discover
+// implementation to inform manager about changes in found resources, e.g. last name of resource
+// "color.example.com/red" is "red".
+type ResourceLastNamesList []string
```
Comment (mpolednik) [inline pkg/dpm/lister.go]: This is a weird name. Having something like DevicePluginList is nicer and more fitting to what this really is.

File: pkg/dpm/manager.go [hunk 2]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+	defer fsWatcher.Close()
 	fsWatcher.Add(pluginapi.DevicePluginPath)
 
-	dpm := &DevicePluginManager{
-		stopCh:    stopCh,
-		signalCh:  sigs,
-		fsWatcher: fsWatcher,
+	// Create list of running plugins and start Discover method of given lister. This method is
+	// responsible of notifying manager about changes in available plugins.
+	var pluginsMap = make(map[string]devicePlugin)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: nit: pluginMap

File: pkg/dpm/manager.go [hunk 3]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+		}()
+	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to start the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) startPlugins() {
-	for _, plugin := range dpm.plugins {
-		go plugin.Start()
+func startUpPlugin(pluginLastName string, plugin devicePlugin) {
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: startPlugin

File: pkg/dpm/manager.go [hunk 4]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+		}()
+	}
+	wg.Wait()
 }
 
-// Run starts the DevicePluginManager.
-func (dpm *DevicePluginManager) Run() {
-	defer dpm.fsWatcher.Close()
-	dpm.startPlugins()
+func (dpm *Manager) shutDownPlugins(pluginsMap map[string]devicePlugin) {
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: stopPlugins

File: pkg/dpm/manager.go [hunk 5]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+		}
 	}
+	startPluginServer(pluginLastName, plugin)
 }
 
-// startPlugins is helper function to stop the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) stopPlugins() {
-	for _, plugin := range dpm.plugins {
-		plugin.Stop()
+func shutDownPlugin(pluginLastName string, plugin devicePlugin) {
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: stopPlugin

File: pkg/dpm/manager.go [hunk 6]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+	}
+	wg.Wait()
 }
 
-// startPlugins is helper function to start the underlying gRPC device plugins.
-func (dpm *DevicePluginManager) startPlugins() {
-	for _, plugin := range dpm.plugins {
-		go plugin.Start()
+func startUpPlugin(pluginLastName string, plugin devicePlugin) {
+	if devicePluginImplementation, ok := plugin.DevicePlugin.(PluginInterfaceStart); ok {
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: devicePluginImpl is slightly more consistent in what we'd normally see in go.

File: pkg/dpm/plugin.go [hunk 7]
```diff
@@ -14,49 +14,102 @@ import (
...
+	DevicePlugin     PluginInterface
+	ResourceName     string
+	ResourceLastName string
+	Socket           string
+	Server           *grpc.Server
+	Running          bool
+	Starting         sync.Mutex
+}
+
+func newDevicePlugin(resourceNamespace string, resourceLastName string, devicePluginImplementation PluginInterface) devicePlugin {
```
Comment (mpolednik) [inline pkg/dpm/plugin.go]: Same comment about devicePluginImpl.

File: pkg/dpm/manager.go [hunk 8]
```diff
@@ -3,99 +3,208 @@ package dpm
...
-	go dpm.handleSignals()
+	// Add new plugins
+	for newPluginLastName, _ := range newPluginsSet {
+		wg.Add(1)
+		go func() {
+			if _, ok := currentPluginsMap[newPluginLastName]; !ok {
+				glog.V(3).Infof("Adding a new plugin \"%s\"", newPluginLastName)
+				plugin := newDevicePlugin(dpm.lister.GetResourceNamespace(), newPluginLastName, dpm.lister.NewPlugin(newPluginLastName))
+				startUpPlugin(newPluginLastName, plugin)
+				currentPluginsMap[newPluginLastName] = plugin
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: We need a mutex around this write, otherwise concurrent map writes might occur.

File: pkg/dpm/manager.go [hunk 9]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+	}
+	wg.Wait()
+
+	// Remove old plugins
+	for pluginLastName, plugin := range currentPluginsMap {
+		wg.Add(1)
+		go func() {
+			if _, ok := newPluginsSet[pluginLastName]; !ok {
+				shutDownPlugin(pluginLastName, plugin)
+				delete(currentPluginsMap, pluginLastName)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: Needs to be guarded by mutex.

File: pkg/dpm/manager.go [hunk 10]
```diff
@@ -3,99 +3,208 @@ package dpm
...
+}
 
-	return dpm
+func (dpm *Manager) stopPluginServers(pluginsMap map[string]devicePlugin) {
+	var wg sync.WaitGroup
+
+	for pluginLastName, plugin := range pluginsMap {
+		wg.Add(1)
+		go func() {
+			stopPluginServer(pluginLastName, plugin)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: wg.Done()

**Review by mpolednik — CHANGES_REQUESTED** [review]
Overall looks great!

Several names (mostly private, one exported) could be improved and we need more robust concurrency guards.

File: pkg/dpm/manager.go (hunk 10 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/plugin.go (hunk 7 above)
Comment (phoracek) [reply pkg/dpm/plugin.go]: Done

File: pkg/dpm/manager.go (hunk 6 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 5 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 4 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 3 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 2 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 9 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 8 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/lister.go (hunk 1 above)
Comment (phoracek) [reply pkg/dpm/lister.go]: Done

**Comment by phoracek** [comment]
Thanks for comments, should be fixed, tried with bridge DP and it worked.

File: pkg/dpm/manager.go [hunk 11]
```diff
@@ -3,99 +3,217 @@ package dpm
...
-	for _, plugin := range dpm.plugins {
-		plugin.Stop()
+func startPlugin(pluginLastName string, plugin devicePlugin) {
+	if devicePluginImpl, ok := plugin.DevicePlugin.(PluginInterfaceStart); ok {
+		err := devicePluginImpl.Start()
+		if err != nil {
+			glog.Errorf("Failed to start plugin \"%s\": %s", pluginLastName, err)
+		}
 	}
+	startPluginServer(pluginLastName, plugin)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: Should the pluginServer start even if the plugin itself failed to start?

File: pkg/dpm/manager.go [hunk 12]
```diff
@@ -3,99 +3,217 @@ package dpm
...
+	}
+}
+
+func startPluginServer(pluginLastName string, plugin devicePlugin) {
+	for i := 1; i <= startPluginServerRetries; i++ {
+		err := plugin.StartServer()
+		if err == nil {
+			return
+		} else if i == startPluginServerRetries {
+			glog.V(3).Infof("Failed to start plugin server \"%s\", within given tries: %s", pluginLastName, err)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: Would be nice to log # of tries (since it'll have to be configurable).

File: pkg/dpm/manager.go [hunk 13]
```diff
@@ -3,99 +3,217 @@ package dpm
...
+			glog.Errorf("Failed to start plugin server \"%s\", waiting %d before next try: %s", pluginLastName, startPluginServerRetryWait, err)
+			time.Sleep(startPluginServerRetryWait)
 		}
 	}
 }
+
+func stopPluginServer(pluginLastName string, plugin devicePlugin) {
+	err := plugin.StopServer()
+	if err != nil {
+		glog.Errorf("Failed to stop plugin \"%s\": %s", pluginLastName, err)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: failed to stop plugin's $name server?

File: pkg/dpm/manager.go [hunk 14]
```diff
@@ -3,99 +3,217 @@ package dpm
...
+
+func startPluginServer(pluginLastName string, plugin devicePlugin) {
+	for i := 1; i <= startPluginServerRetries; i++ {
+		err := plugin.StartServer()
+		if err == nil {
+			return
+		} else if i == startPluginServerRetries {
+			glog.V(3).Infof("Failed to start plugin server \"%s\", within given tries: %s", pluginLastName, err)
+		} else {
+			glog.Errorf("Failed to start plugin server \"%s\", waiting %d before next try: %s", pluginLastName, startPluginServerRetryWait, err)
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: attempt %d out of %d (i, startPluginServerRetries)

File: pkg/dpm/manager.go [hunk 15]
```diff
@@ -3,99 +3,217 @@ package dpm
...
-// DevicePluginManager is container for 1 or more DevicePlugins.
-type DevicePluginManager struct {
-	plugins   []DevicePluginInterface
-	lister    DeviceLister
-	stopCh    chan struct{}
-	signalCh  chan os.Signal
-	fsWatcher *fsnotify.Watcher
+const (
+	startPluginServerRetries   = 3
+	startPluginServerRetryWait = 3 * time.Second
```
Comment (mpolednik) [inline pkg/dpm/manager.go]: Add TODO - make these values configurable.

File: pkg/dpm/plugin.go [hunk 16]
```diff
@@ -14,49 +14,102 @@ import (
...
-	Server       *grpc.Server
-	ResourceName string
-	Deps         DevicePluginInterface
-	Update       chan Message
-	Running      bool
-	Starting     sync.Mutex
+type devicePlugin struct {
+	DevicePlugin     PluginInterface
+	ResourceName     string
+	ResourceLastName string
```
Comment (mpolednik) [inline pkg/dpm/plugin.go]: s/ResourceLastName/Name (to be consistent with what lister provides).

File: pkg/dpm/plugin.go [hunk 17]
```diff
@@ -14,49 +14,102 @@ import (
...
-	Devs         []*pluginapi.Device
-	Socket       string
-	Server       *grpc.Server
-	ResourceName string
-	Deps         DevicePluginInterface
-	Update       chan Message
-	Running      bool
-	Starting     sync.Mutex
+type devicePlugin struct {
+	DevicePlugin     PluginInterface
```
Comment (mpolednik) [inline pkg/dpm/plugin.go]: s/DevicePlugin/DevicePluginImpl

**Review by mpolednik — CHANGES_REQUESTED** [review]
Almost good to go, I've managed to find several small nits.

File: pkg/dpm/plugin.go (hunk 17 above)
Comment (phoracek) [reply pkg/dpm/plugin.go]: Done

File: pkg/dpm/manager.go (hunk 11 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 15 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/plugin.go (hunk 16 above)
Comment (phoracek) [reply pkg/dpm/plugin.go]: Done

File: pkg/dpm/manager.go (hunk 13 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 14 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

File: pkg/dpm/manager.go (hunk 12 above)
Comment (phoracek) [reply pkg/dpm/manager.go]: Done

**Review by mpolednik — APPROVED** [review]

**Pull request merged**